url70997c83aa_Xilaew_risiko_tgz-pJ8-risiko_XMI2XMLtransformerJ8/src/risiko/main/Client2.java

```diff
@@ -0,0 +1,200 @@
+package risiko.main;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.SocketChannel;
+import java.util.Iterator;
+
+import org.eclipse.emf.common.util.EList;
+import org.eclipse.emf.ecore.EObject;
+
+import risiko.GameMonitor;
+import risiko.actions.AddPlayer;
+import risiko.actions.StartGame;
+import risiko.actions.actionFactory;
+import risiko.gamestate.Player;
+import risiko.gamestate.stateFactory;
+
+public class Client2 implements Runnable {
+	private Selector selector;
+	SocketChannel s;
+	private ByteBuffer buf = ByteBuffer.allocate(256);
+	private GameMonitor game = new GameMonitor();
+	private static final int PORT = 10523;
+	private boolean sent = false;
+	private boolean started = false;
+
+	public static void main(String[] args) {
+		SocketAddress addr = new InetSocketAddress("localhost", PORT);
+		try {
+			new Client2(addr).run();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+
+	Client2(SocketAddress address) throws IOException {
+		this.selector = Selector.open();
+		this.s = SocketChannel.open(address);
+		s.configureBlocking(false);
+		ClientState cs = new ClientState("Server");
+		this.s.register(this.selector, SelectionKey.OP_READ, cs);
+	}
+
+	private void handleRead(SelectionKey key) {
+		SocketChannel ch = (SocketChannel) key.channel();
+		ClientState cs = (ClientState) key.attachment();
+		StringBuilder sb = new StringBuilder();
+
+		sb.append(cs.getBuffer());
+
+		buf.clear();
+		int read = 0;
+		try {
+			while ((read = ch.read(buf)) > 0) {
+				buf.flip();
+				byte[] bytes = new byte[buf.limit()];
+				buf.get(bytes);
+				sb.append(new String(bytes));
+				buf.clear();
+			}
+
+			String msg = sb.toString();
+			int i;
+			String current;
+			while ((i = msg.indexOf('\0')) >= 0) {
+				current = msg.substring(0, i);
+				msg = msg.substring(i + 1);
+				try {
+					System.out.println(cs.getName() + ":" + current);
+					// System.out.flush();
+
+					ByteArrayInputStream in = new ByteArrayInputStream(
+							current.getBytes());
+					game.parseAndHandle(in);
+
+				} catch (Exception e) {
+					e.printStackTrace();
+					System.err.flush();
+				}
+			}
+			if (read < 0) {
+				System.out.println(cs.getName() + " left the chat.\n");
+				ch.close();
+			} else {
+				cs.setBuffer(msg);
+			}
+		} catch (IOException e) {
+			key.cancel();
+			e.printStackTrace();
+		}
+	}
+
+	@Override
+	public void run() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.